src/QsCompiler/Transformations/TypeResolutionSummary.cs

```diff
@@ -0,0 +1,390 @@
+
+namespace Microsoft.Quantum.QsCompiler
+{
+    using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
+    using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    // Type Parameters are frequently referenced by the callable of the type parameter followed by the name of the specific type parameter.
+    using TypeParameterName = Tuple<QsQualifiedName, NonNullable<string>>;
+    using TypeParameterResolutions = ImmutableDictionary</*TypeParameterName*/ Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>;
+
+    /// <summary>
+    /// Summarizes a series of type parameter resolution dictionaries, ResolutionsInSummary, into one
+    /// resolution dictionary, CombinedTypeParameterResolutions, containing the ultimate type resolutions
+    /// for all the type parameters found in the dictionaries. Validation is done on the resolutions,
+    /// which can be checked through the IsValidSummary, SummarizesOverConflictingResolution, and
+    /// SummarizesOverParameterConstriction flags.
+    /// </summary>
+    public class TypeResolutionSummary
+    {
+        // Static Members
+
+        /// <summary>
+        /// Checks if the given type parameter directly resolves to itself. If so, the
+        /// resolution may be overridden without conflict.
+        /// </summary>
+        private static bool IsSelfResolution(TypeParameterName typeParam, ResolvedType res)
+        {
+            return res.Resolution is ResolvedTypeKind.TypeParameter tp
+                && tp.Item.Origin.Equals(typeParam.Item1)
+                && tp.Item.TypeName.Equals(typeParam.Item2);
+        }
+
+        /// <summary>
+        /// Reverses the dependencies of type parameters resolving to other type parameters in the given
+        /// dictionary to create a lookup whose keys are type parameters and whose values are all the type
+        /// parameters that can be updated by knowing the resolution of the lookup's associated key.
+        /// </summary>
+        private static ILookup<TypeParameterName, TypeParameterName> GetReplaceable(TypeParameterResolutions.Builder typeParamResolutions)
+        {
+            return typeParamResolutions
+               .Select(kvp => (kvp.Key, GetTypeParameters.Apply(kvp.Value))) // Get any type parameters in the resolution type.
+               .SelectMany(tup => tup.Item2.Select(value => (tup.Key, value))) // For each type parameter found, match it to the dictionary key.
+               .ToLookup(// Reverse the keys and resulting type parameters to make the lookup.
+                   kvp => kvp.value,
+                   kvp => kvp.Key);
+        }
+
+        // Fields and Properties
+
+        /// <summary>
+        /// Array of all the type parameter resolution dictionaries that are summarized in this summary.
+        /// </summary>
+        public readonly ImmutableArray<TypeParameterResolutions> ResolutionsInSummary;
```

Comment:
I liked the previous name that we arrived to in the previous PR. Something similar to `IndependentResolutionsDictionary` would be great to convey what this field is about.

Reply:
I renamed it to `IndependentResolutionDictionaries` and also renamed `CombinedTypeParameterResolutions` to `CombinedResolutionDictionary` to match it.